quill-core/src/main/scala/io/getquill/dsl/DynamicQueryDSL.scala

```diff
@@ -0,0 +1,376 @@
+      case None    => DynamicSetEmpty()
+    }
+
+  def set[T, U](property: String, value: Quoted[U]): DynamicSet[T, U] =
+    set((f: Quoted[T]) => splice(Constant(property)), value)
+
+  def setValue[T, U](property: String, value: U)(implicit enc: Encoder[U]): DynamicSet[T, U] =
+    set(property, spliceLift(value))
+
+  def dynamicQuerySchema[T](entity: String, columns: DynamicAlias[T]*): DynamicEntityQuery[T] = {
+    val aliases =
+      columns.map { alias =>
+
+        @tailrec def path(ast: Ast, acc: List[String] = Nil): List[String] =
+          ast match {
+            case Property(a, name) =>
+              path(a, name :: acc)
+            case _ =>
+              acc
+          }
+
+        PropertyAlias(path(alias.property(splice[T](Ident("v"))).ast), alias.name)
+      }
+    DynamicEntityQuery(splice[EntityQuery[T]](Entity(entity, aliases.toList)))
+  }
+
+  private[this] val nextIdentId = new DynamicVariable(0)
+
+  private[this] def withFreshIdent[R](f: Ident => R): R = {
+    val idx = nextIdentId.value
+    nextIdentId.withValue(idx + 1) {
+      f(Ident(s"v$idx"))
+    }
+  }
+
+  private def dyn[T](ast: Ast): DynamicQuery[T] =
+    DynamicQuery[T](splice[Query[T]](ast))
+
+  private def splice[T](a: Ast) =
+    new Quoted[T] {
+      override def ast = a
+    }
+
+  protected def spliceLift[O](o: O)(implicit enc: Encoder[O]) =
+    splice[O](ScalarValueLift("o", o, enc))
+
+  object DynamicQuery {
+    def apply[T](p: Quoted[Query[T]]) =
+      new DynamicQuery[T] {
+        override def q = p
+      }
+  }
+
+  sealed trait DynamicQuery[+T] {
+
+    protected[getquill] def q: Quoted[Query[T]]
+
+    protected[this] def transform[U, V, R](f: Quoted[U] => Quoted[V], t: (Ast, Ident, Ast) => Ast, r: Ast => R = dyn _) =
+      withFreshIdent { v =>
+        r(t(q.ast, v, f(splice(v)).ast))
+      }
+
+    protected[this] def transformOpt[O, R, D <: DynamicQuery[T]](opt: Option[O], f: (Quoted[T], Quoted[O]) => Quoted[R], t: (Quoted[T] => Quoted[R]) => D, thiz: D)(implicit enc: Encoder[O]) =
+      opt match {
+        case Some(o) =>
+          t(v => f(v, spliceLift(o)))
+        case None =>
+          thiz
+      }
+
+    def map[R](f: Quoted[T] => Quoted[R]): DynamicQuery[R] =
+      transform(f, Map)
+
+    def flatMap[R](f: Quoted[T] => Quoted[Query[R]]): DynamicQuery[R] =
+      transform(f, FlatMap)
+
+    def filter(f: Quoted[T] => Quoted[Boolean]): DynamicQuery[T] =
+      transform(f, Filter)
+
+    def withFilter(f: Quoted[T] => Quoted[Boolean]): DynamicQuery[T] =
+      filter(f)
+
+    def filterOpt[O](opt: Option[O])(f: (Quoted[T], Quoted[O]) => Quoted[Boolean])(implicit enc: Encoder[O]): DynamicQuery[T] =
+      transformOpt(opt, f, filter, this)
+
+    def concatMap[R, U](f: Quoted[T] => Quoted[U])(implicit ev: U => Traversable[R]): DynamicQuery[R] =
+      transform(f, ConcatMap)
+
+    def sortBy[R](f: Quoted[T] => Quoted[R])(implicit ord: OrdDsl#Ord[R]): DynamicQuery[T] =
```

Comment:
Maybe have `sortByOpt` ?

Reply:
I think it's reasonable that the users would need to do something like:
```
if(sort) q.sortBy(..) else q
``